core/store/src/config.rs

```diff
@@ -176,3 +190,74 @@ impl Default for MigrationSnapshot {
         Self::Enabled(true)
     }
 }
+
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+pub struct TrieCacheConfig {
+    /// Limit the number trie nodes that fit in the cache.
+    pub max_entries: Option<u64>,
+    /// Limit the sum of all value sizes that fit in the cache.
+    pub max_bytes: Option<u64>,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct ShardUIdShortSerialization(ShardVersion, ShardId);
+
+impl From<ShardUIdShortSerialization> for ShardUId {
```

Comment:
I would prefer us to define a reasonable Display & FromStr for ShardUid (and improve Debug along the way).

Reply:
Hm, then what would be a reasonable general-purpose string format of ShardUId? One that also works as JSON key for the config?
"shardXvY" doesn't feel right anywhere other than when used as a JSON key.

Reply:
I would say `"shard1v0"` would be better than our current debug of `"ShardUId { version: 0, shard_id: 1}"`. Though, I'd probably go for something more compact like `s2:1`. 

The rule-of-thumb here is that it's useful to have canonical string representations for vocabulary types, as this simplifies debugging quite a bit. 

Don't have a super strong opinion here!

Reply:
That makes sense, I'm on board with using FromStr and Display here.
But I was hoping that I can at least avoid a custom serde serialization / deserialization implementation that way. But [DisplayFromStr](https://docs.rs/serde_with/latestserde_with/struct.DisplayFromStr.html) only works for fields. I think I would need it on the struct to make it work in a map.

Am I missing something, or would your suggested solution end up with custom implementation of `Debub`, `Display`, `FromStr`, `serde::Deserialize`, and `serde::Serialize` for ShardUId?

Reply:
Yeah, we'd need all those custom impls. But we perhaps can come up with some generic way to impl serde in terms of display/from string (I think we already hard-code it a couple of times somewhere). 

Last time I look at it, I didn't figure a super-nice way to do that, so we might go for a less-nice way (https://docs.rs/serde_with/latest/serde_with/, or a macro_rules).

Reply:
>But [DisplayFromStr](https://docs.rs/serde_with/latestserde_with/struct.DisplayFromStr.html) only works for fields.

Ah, I think you are lookin for 

* https://docs.rs/serde_with/latest/serde_with/derive.DeserializeFromStr.html
* https://docs.rs/serde_with/latest/serde_with/derive.SerializeDisplay.html

Reply:
Ah yes, that's what I was looking for thanks! But looking at it again, it seems that a custom implementation is simpler than pulling in another dependency. (I didn't realize it would be an extra dependency at first.)

I could find at least one more place that could also use `DeserializeFromStr` and probably there are many playes. If we want to use `serde_with` we should probably do it in a separate PR and use it in as many places as possible. But I'm not even sure it would be an improvement.